File: forms/framingham-risk-score-for-hard-coronary-heart-disease/front-end-patient-form-with-html/js/risk-grader.js
import { calculateFraminghamRisk, isLikelyDraft } from './utils.js';
import { allRules } from './risk-rules.js';

export function calculateRisk(data) {
  if (isLikelyDraft(data)) return { riskCategory: 'draft', tenYearRiskPercent: 0, firedRules: [] };

  const rawRisk = calculateFraminghamRisk(data);
  const tenYearRiskPercent = Math.round(rawRisk * 10) / 10;

  const firedRules = [];
  for (const rule of allRules()) {
    if (rule.evaluate(data, tenYearRiskPercent)) {
      firedRules.push({ id: rule.id, category: rule.category, description: rule.description, riskLevel: rule.riskLevel });
    }
  }

  let riskCategory;
  if (tenYearRiskPercent < 10) riskCategory = 'low';
  else if (tenYearRiskPercent < 20) riskCategory = 'intermediate';
  else riskCategory = 'high';

  return { riskCategory, tenYearRiskPercent, firedRules };
}

File: forms/systematic-coronary-risk-evaluation-2-diabetes/full-stack-with-rust-axum-loco-tera-htmx-alpine/src/engine/risk_grader.rs
use super::risk_rules::all_rules;
use super::types::{AssessmentData, FiredRule, RiskCategory};
use super::utils::is_likely_draft;

/// Pure function: evaluates all risk rules against patient data.
/// Returns risk category, and fired rules.
/// Risk category is the maximum among all fired rules.
pub fn calculate_risk(data: &AssessmentData) -> (RiskCategory, Vec<FiredRule>) {
    if is_likely_draft(data) {
        return ("draft".to_string(), vec![]);
    }

    let mut fired_rules = Vec::new();

    for rule in all_rules() {
        if (rule.evaluate)(data) {
            fired_rules.push(FiredRule {
                id: rule.id.to_string(),
                category: rule.category.to_string(),
                description: rule.description.to_string(),
                risk_level: rule.risk_level.to_string(),
            });
        }
    }

    let risk_order = |level: &str| -> u8 {
        match level {
            "low" => 1,
            "medium" => 2,
            "high" => 3,
            _ => 0,
        }
    };

    let risk_category = if fired_rules.is_empty() {
        "low".to_string()
    } else {
        let max_level = fired_rules
            .iter()
            .map(|r| r.risk_level.as_str())
            .max_by_key(|l| risk_order(l))
            .unwrap_or("low");
        // Map to SCORE2-Diabetes categories
        match max_level {
            "high" => "veryHigh".to_string(),
            "medium" => "high".to_string(),
            _ => "moderate".to_string(),
        }
    };

    (risk_category, fired_rules)
}
